### PR TITLE
Improve handling of search for YTDL executables

### PR DESCRIPTION
**Pull request by techanon**
- Improve path search logic to include some standard YTDL (not just yt-dlp) executable names.  
- Add menu option for specifying a precise custom executable location under Tools -> AudioLink -> Select - Custom YTDL Location.  
- Move _localytdlpPath use to windows only as it's an exe specific path.  
- Update LocateExecutable to support intaking a list of possible known ytdl executable names to improve search reliability.  
- Update LocateExecutable to support PATH quirks with certain MSeries Macbooks.

### CODE REVIEW

**Comment by float3** [comment]
under what scenarios?

**Comment by float3** [comment]
I don't remember what format ytdlp serves when the source is youtube, but maybe there's a way to get it to serve a specific format? if that was the case I would be okay with having it available on linux

**Comment by float3** [comment]
I'm not sure if we want to support ytdl

**Comment by float3** [comment]
> So it's an editor specific thing then. Ok.

did you test your changes on mac in editor?

**Comment by techanon** [comment]
> did you test your changes on mac in editor?

That specific Mac part is ported from the changes I made to VideoPlayerShim based on the feedback from an M-Series Mac user. The shim uses somewhat different search logic, but the path requirement for m-series is the same.

> I'm not sure if we want to support ytdl

Is there a reason why not? Also it's all local side. Isn't it better to support the canonical name of ytdl as well as the ytdlp fork when searching for a users local install? The search logic prioritizes VRChat's official install first anyways, as well as ytdlp being searched for first in the list of possible names.

File: Packages/com.llealloo.audiolink/Runtime/Scripts/ytdlpPlayer.cs [hunk 1]
```diff
@@ -155,36 +172,43 @@ public static bool IsytdlpAvailable()
...
+            if (!string.IsNullOrEmpty(customPath))
+            {
+                if (File.Exists(customPath))
+                {
+                    Debug.Log($"[AudioLink:ytdlp] Custom YTDL location found: {customPath}");
+                    _ytdlpPath = customPath;
+                    _ytdlpFound = true;
+                }
+
+                Debug.LogWarning($"[AudioLink:ytdlp] Custom YTDL location detected but does not exist: {customPath}");
```
Comment (pema99) [inline Packages/com.llealloo.audiolink/Runtime/Scripts/ytdlpPlayer.cs:187]: This doesn't seem right? There should be no warning printed if we found the executable in the branch above, right?

File: Packages/com.llealloo.audiolink/Runtime/Scripts/ytdlpPlayer.cs [hunk 2]
```diff
@@ -143,6 +141,25 @@ public static class ytdlpURLResolver
...
         private static bool _ytdlpFound = false;
 
+        private const string userDefinedYTDLPathKey = "YTDL-PATH-CUSTOM";
+        private const string userDefinedYTDLPathMenu = "Tools/AudioLink/Select Custom YTDL Location";
+
+        [MenuItem(userDefinedYTDLPathMenu, priority = 1)]
+        private static void SelectYtdlInstall()
+        {
+            string ytdlPath = EditorPrefs.GetString(userDefinedYTDLPathKey, "");
+            string tpath = ytdlPath.Substring(0, ytdlPath.LastIndexOf("/", StringComparison.Ordinal) + 1);
```
Comment (pema99) [inline Packages/com.llealloo.audiolink/Runtime/Scripts/ytdlpPlayer.cs:151]: Won't this break if you use `\` as the path separator?

File: Packages/com.llealloo.audiolink/Runtime/Scripts/ytdlpPlayer.cs [hunk 3]
```diff
@@ -301,7 +345,9 @@ public override void OnInspectorGUI()
                     if (EditorGUI.EndChangeCheck())
                     {
                         EditorUtility.SetDirty(_ytdlpPlayer);
-                    };
+                    }
+
+                    ;
```
Comment (pema99) [inline Packages/com.llealloo.audiolink/Runtime/Scripts/ytdlpPlayer.cs:350]: typo looks like

**Comment by pema99** [comment]
Looks sane overall. Perhaps there should be a way to clear the custom path you've set previously though?

File: Packages/com.llealloo.audiolink/Runtime/Scripts/ytdlpPlayer.cs (hunk 1 above)
Comment (techanon) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/ytdlpPlayer.cs:187]: Return statement is missing. Will fix.

File: Packages/com.llealloo.audiolink/Runtime/Scripts/ytdlpPlayer.cs (hunk 2 above)
Comment (techanon) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/ytdlpPlayer.cs:151]: I did try using the `\` separator check but what I saw was that Unity was normalizing to `/` for the paths.

**Comment by techanon** [comment]
The current way that the custom path gets cleared is by clicking the menu option to open the file select dialog, and then clicking cancel. This returns an empty string path which is detected as not provided, thus falling back on the auto-detect methods.

**Comment by float3** [comment]
> The current way that the custom path gets cleared is by clicking the menu option to open the file select dialog, and then clicking cancel. This returns an empty string path which is detected as not provided, thus falling back on the auto-detect methods.

yeah this is not an acceptable method of clearing the custom path

**Comment by techanon** [comment]
We can switch it to a menu toggle style. When active, selecting it again implicitly clears it, then to pick a new path, select the menu again.
How's that sound?

**Comment by pema99** [comment]
Sounds like an improvement to me

**Comment by techanon** [comment]
Meant to push to a separate branch.

**Comment by pema99** [comment]
Got some conflicts now, probably with your other PR

**Comment by techanon** [comment]
@pema99 fixed.

**Pull request merged**